include/seqan3/search/algorithm/search_result_range.hpp

```diff
@@ -0,0 +1,324 @@
+    using single_pass_query_range_t = decltype(views::zip(std::views::iota(0), std::declval<query_range_t &&>())
+                                             | views::single_pass_input);
+    //!\brief The iterator over the wrapped query range.
+    using single_pass_query_range_iterator = std::ranges::iterator_t<single_pass_query_range_t>;
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    search_result_range() = default; //!< Defaulted.
+    search_result_range(search_result_range const &) = delete; //!< This is a move-only type.
+    search_result_range(search_result_range &&) = default; //!< Defaulted.
+    search_result_range & operator=(search_result_range const &) = delete; //!< This is a move-only type.
+    search_result_range & operator=(search_result_range &&) = default; //!< Defaulted.
+    ~search_result_range() = default; //!< Defaulted.
+
+    /*!\brief Constructs a search result range.
+     *
+     * \param[in] search_algorithm The search algorithm to invoke on a single query.
+     * \param[in] query_range The range over the queries to be searched.
+     */
+    search_result_range(search_algorithm_t search_algorithm, query_range_t query_range) :
+        search_algorithm{std::move(search_algorithm)},
+        single_pass_query_range{views::zip(std::views::iota(0), std::move(query_range))},
+        next_query_it{single_pass_query_range.begin()}
+    {}
+    //!\}
+
+    /*!\name Iterators
+     * \{
+     */
+
+    /*!\brief Returns an iterator to the first element of the search result range.
+     * \return An iterator to the first element.
+     *
+     * \details
+     *
+     * The fist invocation of this function will invoke the search for the first query in the underlying range.
+     */
+    constexpr search_result_range_iterator begin()
+    {
+        search_result_range_iterator tmp_it{*this, first_invocation_of_begin};
+        first_invocation_of_begin = true;
+        return tmp_it;
+    }
+
+    search_result_range_iterator begin() const = delete;
+    search_result_range_iterator cbegin() const = delete;
+
+    /*!\brief Returns a sentinel signaling the end of the search result range.
+     * \return a sentinel.
+     *
+     * \details
+     *
+     * The search result range is an input range and the end is reached when all results have been processed.
+     */
+    constexpr std::ranges::default_sentinel_t end() noexcept
+    {
+        return {};
+    }
+
+    constexpr std::ranges::default_sentinel_t end() const = delete;
+    constexpr std::ranges::default_sentinel_t cend() const = delete;
+    //!\}
+
+protected:
+    /*!\brief Receives the search results from the next query.
+     *
+     * \returns `true` if hits could be found for the current query, otherwise `false`.
+     *
+     * \details
+     *
+     * Invokes the search on the current query and increments the query iterator.
+     */
+    bool next()
```

Comment:
Personal preference. There is no advantage in making it private either. It would allow you to inherit from this and use the functionality to do something. But I agree that this might not be ever the case.